kafka/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/AzureAuthenticationCallbackHandler.scala

```diff
@@ -0,0 +1,85 @@
+/**
+  * Copyright (c) 2013-present Snowplow Analytics Ltd.
+  * All rights reserved.
+  *
+  * This program is licensed to you under the Snowplow Community License Version 1.0,
+  * and you may not use this file except in compliance with the Snowplow Community License Version 1.0.
+  * You may obtain a copy of the Snowplow Community License Version 1.0 at https://docs.snowplow.io/community-license-1.0
+  */
+package com.snowplowanalytics.snowplow.collectors.scalastream
+package sinks
+
+import java.net.URI
+import java.{lang, util}
+
+import javax.security.auth.callback.Callback
+import javax.security.auth.callback.UnsupportedCallbackException
+import javax.security.auth.login.AppConfigurationEntry
+
+import org.apache.kafka.clients.producer.ProducerConfig
+import org.apache.kafka.common.security.auth.AuthenticateCallbackHandler
+import org.apache.kafka.common.security.oauthbearer.OAuthBearerToken
+import org.apache.kafka.common.security.oauthbearer.OAuthBearerTokenCallback
+
+import com.microsoft.azure.credentials.MSICredentials
+
+import com.nimbusds.jwt.JWTParser
+
+class AzureAuthenticationCallbackHandler extends AuthenticateCallbackHandler {
+
+  val credentials: MSICredentials = {
+    val clientId = sys.env.get("AZURE_CLIENT_ID_FOR_EVENT_HUB").orElse(sys.env.get("AZURE_CLIENT_ID"))
+    val creds    = new MSICredentials()
+    clientId.map(creds.withClientId).getOrElse(creds)
+  }
+
+  var sbUri: String = ""
+
+  override def configure(
+    configs: util.Map[String, _],
+    saslMechanism: String,
+    jaasConfigEntries: util.List[AppConfigurationEntry]
+  ): Unit = {
+    val bootstrapServer =
+      configs
+        .get(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG)
+        .toString
+        .replaceAll("\\[|\\]", "")
+        .split(",")
+        .toList
+        .headOption match {
+        case Some(s) => s
+        case None    => throw new Exception("Empty bootstrap servers list")
+      }
+    val uri = URI.create("https://" + bootstrapServer)
```

Comment:
I am curious why we set the auth uri against the eventhubs bootstrap server from the producer conf? The managed identity / service account is at the active directory level so it might need to be updated use https://login.microsoftonline.com/ or possibly https://management.core.windows.net/.  There's a [powershell](https://github.com/logcorner/Getting-an-access-token-inside-an-aks-pod-using-Azure-Managed-Identities-with-Workload-Identity-Fede/blob/main/my-script.ps1) example of what I believe we're trying to achieve.

For example, we deploy a [service account](https://github.com/snowplow-devops/helm-charts/blob/main/charts/cloudserviceaccount/templates/serviceaccount.yaml#L13) in AKS which adds the following ENV variables on the collector pod

```
AZURE_AUTHORITY_HOST: https://login.microsoftonline.com/
AZURE_CLIENT_ID: aabb44cc69f-e118-4636-b21b-f43598fdkj39fh33jdjfk   
AZURE_FEDERATED_TOKEN_FILE: /var/run/secrets/azure/tokens/azure-identity-token
AZURE_TENANT_ID: 333333-b20a-4bda-ade0-3423432443
```

The AZURE_AUTHORITY_HOST is `https://login.microsoftonline.com/` so I would've expected this to be what we would auth against for a token.

The `AZURE_CLIENT_ID` and `AZURE_FEDERATED_TOKEN_FILE` are both important here as `AZURE_CLIENT_ID` is the managed identity ID and `AZURE_FEDERATED_TOKEN_FILE` contains  token for the service account 

```
{
  "aud": [
    "api://AzureADTokenExchange"
  ],
  "exp": 1703283977,
  "iat": 1703280377,
  "iss": "https://eastus.oic.prod-aks.azure.com/d21c78e8-a114-11ee-8c90-0242ac120002
/d8519c0c-a114-11ee-8c90-0242ac120002/",
  "kubernetes.io": {
    "namespace": "collector-namespace",
    "pod": {
      "name": "my-collector-pod-5f8f495dc7-rnk5h",
      "uid": "ee25c8d2-a114-11ee-8c90-0242ac120002"
    },
    "serviceaccount": {
      "name": "collector-serice-account",
      "uid": "fbb17f6e-a114-11ee-8c90-0242ac120002"
    }
  },
  "nbf": 1703280377,
  "sub": "system:serviceaccount:collector-namespace:collector-service-account"
}
```

With these values using azure cli we can generate a token like 

```
az login --federated-token "$(cat $AZURE_FEDERATED_TOKEN_FILE)" --service-principal -u $AZURE_CLIENT_ID -t $AZURE_TENANT_ID
```

and then run the needed az commands to interact with whatever azure resources have been granted access to the managed identity.

Reply:
We are using `uri` with [MSICredentials.getToken](https://github.com/Azure/autorest-clientruntime-for-java/blob/master/azure-client-authentication/src/main/java/com/microsoft/azure/credentials/MSICredentials.java) method in `getOAuthBearerToken`. According to my understanding, it is more like a resource id than auth uri. We are using bootstrap server address as resource name as [documented here](https://learn.microsoft.com/en-us/azure/event-hubs/authenticate-application).

However, the problem we are facing at the moment I think is using `MSICredentials`. Since we want to deploy Collector in AKS, we have to use Workload Identity auth however it isn't possible with `MSICredentials`. If we use [DefaultAzureCredential](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredential.java) to get token, it should be possible to use Workload Identity auth as well.